src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+{
+	DWORD available;
+
+	return poll->fd != -1 && (!PeekNamedPipe((HANDLE) _get_osfhandle(poll->fd), NULL, 0, NULL,
+		&available, NULL) || available);
+}
+
+
+static gboolean spawn_source_check(GSource *source)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+
+	return (sp->std[IN].fd != -1 && GetTickCount() - sp->last_send_ticks >= 50) ||
+		spawn_source_peek_pipe(&sp->std[OUT]) || spawn_source_peek_pipe(&sp->std[ERR]);
+}
+#else  /* G_OS_WIN32 */
+static gboolean spawn_source_check(GSource *source)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+
+	return (sp->std[IN].fd != -1 && sp->std[IN].revents) ||
+		(sp->std[OUT].fd != -1 && sp->std[OUT].revents) ||
+		(sp->std[ERR].fd != -1 && sp->std[ERR].revents);
+}
+#endif  /* G_OS_WIN32 */
+
+
+static gboolean spawn_source_dispatch(GSource *source, G_GNUC_UNUSED GSourceFunc callback,
+	G_GNUC_UNUSED gpointer user_data)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+	gchar *line;
+	int i;
+#ifdef G_OS_WIN32
+	DWORD status = STILL_ACTIVE;
+#else
+	int status = -1;
+#endif
+
+	if (sp->std[IN].fd != -1)
+	{
+		ssize_t count = write(sp->std[IN].fd, sp->stdin_pos, sp->stdin_remaining);
+
+		if (count > 0)
+		{
+			sp->stdin_remaining -= count;
+
+			if (sp->stdin_remaining)
+			{
+				sp->stdin_pos += count;
+			#ifdef G_OS_WIN32
+				sp->last_send_ticks = GetTickCount();
+			#endif
+			}
+			else
+			{
+				close(sp->std[IN].fd);
+				sp->std[IN].fd = -1;
+				g_free(sp->stdin_text);
+				sp->stdin_text = NULL;
+
+				if (sp->std[OUT].fd == -1 && sp->std[ERR].fd == -1)
+				{
+					g_source_ref(source);
+					g_child_watch_add(sp->pid, spawn_source_exit_cb, sp);
+					return FALSE;
+				}
+
+			#ifdef G_OS_UNIX
+				g_source_remove_poll(source, &sp->std[IN]);
+			#endif
+			}
+		}
+	}
+
+	for (i = 0; i < 2; i++)
+	{
+		while (sp->std[i].fd != -1 && g_io_channel_read_line(sp->ioc[i], &line, NULL,
+			NULL, NULL) == G_IO_STATUS_NORMAL && line)
+		{
+			if (sp->io_cb[i])
+				sp->io_cb[i](line, sp->io_data[i]);
+
+			g_free(line);
+		}
+	}
+
+#ifdef G_OS_WIN32
+	if (!GetExitCodeProcess(sp->pid, &status) || status != STILL_ACTIVE)
+#else
+	if (waitpid(sp->pid, &status, WNOHANG))
+#endif
+	{
+		if (sp->exit_cb)
+			sp->exit_cb(status, sp->exit_data);
+
+		if (sp->dialog)
+			gtk_widget_destroy(sp->dialog);
+
+		g_free(sp->stdin_text);
+		g_spawn_close_pid(sp->pid);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+
+static void spawn_source_finalize(G_GNUC_UNUSED GSource *source)
+{
+}
+
+
+static GSourceFuncs spawn_source_funcs = { spawn_source_prepare, spawn_source_check,
+	spawn_source_dispatch, spawn_source_finalize, NULL, NULL };
+
+
+static void spawn_dialog_response(GtkDialog *dialog, G_GNUC_UNUSED gint response_id,
+	gpointer user_data)
+{
+	SpawnSource *sp = (SpawnSource *) user_data;
+
+	gtk_message_dialog_format_secondary_text(GTK_MESSAGE_DIALOG(sp->dialog),
+		_("Stopping %s..."), sp->command);
+	spawn_kill_process(sp->pid, NULL);
+	g_signal_stop_emission_by_name(dialog, "response");

```

Comment:
> The signal should be stopped, because otherwise it'll close the dialog, and I don't want that before the program terminates (which may not be immediate). 

Oooh, ok, my bad.  Indeed, makes sense.  For some crazy reason I thought this was blocking a second call to this, never mind.

> Disabling Cancel is a good idea.

Thinking about it, it might not be so great, what if the program requires killing twice? (e.g. if it gets stuck in its signal handler)  Of course it would be better UI to disable the Cancel button after clicking on it, but if it is useful to kill twice it might be worth keeping this possibility.